comp/trace/agent/agent.go

```diff
@@ -0,0 +1,188 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package agent defines the tracer agent.
+package agent
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/signal"
+	"runtime"
+	"runtime/pprof"
+	"sync"
+	"syscall"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/watchdog"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"go.uber.org/fx"
+
+	"github.com/DataDog/datadog-agent/comp/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/pidfile"
+	pkgagent "github.com/DataDog/datadog-agent/pkg/trace/agent"
+	"github.com/DataDog/datadog-agent/pkg/trace/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const messageAgentDisabled = `trace-agent not enabled. Set the environment variable
+DD_APM_ENABLED=true or add "apm_config.enabled: true" entry
+to your datadog.yaml. Exiting...`
+
+// ErrAgentDisabled indicates that the trace-agent wasn't enabled through environment variable or config.
+var ErrAgentDisabled = errors.New(messageAgentDisabled)
+
+type dependencies struct {
+	fx.In
+
+	Lc         fx.Lifecycle
+	Shutdowner fx.Shutdowner
+
+	Config             config.Component
+	Context            context.Context
+	Params             *Params
+	TelemetryCollector telemetry.TelemetryCollector
+}
+
+type agent struct {
+	*pkgagent.Agent
+
+	cancel             context.CancelFunc
+	config             config.Component
+	ctx                context.Context
+	params             *Params
+	shutdowner         fx.Shutdowner
+	telemetryCollector telemetry.TelemetryCollector
+	wg                 sync.WaitGroup
+}
+
+func newAgent(deps dependencies) Component {
+	c := component{}
+	deps.Lc.Append(fx.Hook{OnStart: c.start, OnStop: c.stop})
```

Comment:
Usually `deps.Lc.Append` is called after the check if the component is enabled.
This would make the code easier because you can capture the instance of `agent` and so don't need to have `Optional` field.
The following code
```go
deps.Lc.Append(fx.Hook{OnStart: c.start, OnStop: c.stop})
tracecfg := deps.Config.Object()
	if !tracecfg.Enabled {
           // .... 
        }
// ....
c.Optional = util.NewOptional[agentComponent](ag)
```
can be replaced by
```go
tracecfg := deps.Config.Object()
	if !tracecfg.Enabled {
           // .... 
        }
// ... 
deps.Lc.Append(fx.Hook{
      OnStart: func(_ context.Context) error { return start(ag)}),
       OnStop: func(_ context.Context) error { return stop(ag)) })
```